### ISSUE TITLE
Extensions Directory Submission Criteria & Validation

### ISSUE THREAD

**Issue by josephfusco** [wp-graphql/wp-graphql#3205]
# WPGraphQL Extensions Submission Policy

We are establishing a formal submission policy for the WPGraphQL Extensions page to ensure that plugins/extensions are well-maintained, tested, and provide adequate support to users. We invite the community to discuss the proposed process and criteria.

## Submission Process

1. Fork the WPGraphQL repository.
2. Create a new branch.
3. Modify the `extensions.json` file: Add your plugin's details (name, description, URLs, and authors).
4. Validate your JSON against the provided `extensions.json.schema`.
5. Submit a Pull Request.

## Criteria for Inclusion

- **Support Link (Required)**: Must provide a clear support channel.
- **Active Maintenance**: Plugins should be regularly maintained.
- **Tests (Recommended)**: Plugins with test coverage are strongly encouraged.
- Plugins inactive for **over 2 years** may be removed.

## Schema and Automation

- We will use a **JSON schema** to validate submissions and ensure consistency.
- **GitHub Actions** will assist in automating the validation process.

## extensions.json (proposed)

```jsonc
{
  "extensions": [
    {
      "name": "Plugin Name",                         // Required: Name of the plugin
      "description": "Short description here",       // Required: Brief description of what the plugin does
      "plugin_url": "https://plugin-url.com",        // Required: URL to the plugin repository or download
      "support_url": "https://support-url.com",      // Required: URL where users can get support
      "authors": [                                   // Required: List of authors/maintainers
        {
          "name": "Author Name",                     // Required: Author or maintainer name
          "homepage": "https://author-homepage.com"  // Required: URL to author's homepage or profile
        }
      ],
      "plugin_path": "plugin-folder/plugin-file.php", // Optional: Path to the plugin's main file
      "settings_path": "options-general.php?page=plugin-settings", // Optional: Path to plugin's settings page
    }
  ]
}
```

### Community Feedback

We’d love to hear your thoughts and suggestions on improving this process! Please share any feedback or ideas for making this more efficient for both developers and users.

**Comment by moonmeister** [comment]
This all seems pretty reasonable. I think the hard part is what happens after plugins are allowed onto the list. The two year cut of seams generous but its a start. I'm not sure if we want to include some kind of official "up to the discretion of the WPGraphQL maintainers" as a catch all. 

I think enforcement becomes the hard part as well, we should probably have a git hook or GH Action that checks the JSON schema for us to make this easier on users and maintainers. Also if we could build an action that compared public git repos for activity against that two year number (or whatever length is decided) that could be helpful.

All this to say pruning and integrity checks over time will be the hard part.

**Comment by justlevine** [comment]
Few thoughts:

### 1 + 3. JSON Flow:
  - Are we using JSON bc long-term we want to serve this via response payload instead of a plugin? Otherwise, using PHP would allow us to just use PHPStan and an array doctype, without all the performance headaches of correctly caching the parsed file.
  - If we're anyway adding new workflows, why not drop the PR requirement altogether, and use a GH issue template that the workflow converts to a commit (LT this could be directly be swapped out with a .com form submission > approval process )
  - Either way I'm assuming there's a mechanism to filter this list locally? (e.g. a WPE mu-plugin for their algolia extension?)
### 2. Criteria for inclusion:
  - Only the first item (a support link) is an actual criterion - and that should probably be clarified as well as what qualifies.
    - The existence of tests is meaningless if the tests arent well written. Coverage isnt a good proxy either. (e.g. wp-graphql-content-blocks)
    - `Active Maintenance` is subjective, and `Plugins inactive for over 2 years` is also meaningless if it's still compatible (e.g. wp-graphql-upload)
  - Not mentioned is whether the plugin is free or where it is hosted. Personally, I recommend making a _public_ repo a requirement (wordpress's svn can count), otherwise we will exacerbate all the review issues @moonmeister highlighted.
  - Future: Once #2543 is implemented, then the requisite version compatibility headers should be a requirement.
  - Is any actual code review going to be part of the process (the proposed active maintenance, and test coverage indicate yes) - if so what's being checked/smoketested should be defined as well.

### 4. `extensions.json`
  - `authors`: recommend only allowing a singular author, otherwise we're absorbing unnecessary maintenance debt.
  - `description` should have a char limit
  - What are the point of `plugin_path` and `setting_path`? It's (likely unnecessary) maintenance debt.
  - Would suggesting adding a `repo_url` in addition to the support url and the plugin (marketing landing page) url.

**Comment by moonmeister** [comment]
All good points from @justlevine. Public URL was kinda assumed in my head but fully agree it should be a thing. Not entirely sure I agree SVN should count but not a hill I'd die on.

**Comment by renatonascalves** [comment]
> Plugins inactive for over 2 years may be removed.

I disagree strongly with this one. Specially for extensions for WP-GraphQL.

**Comment by jasonbahl** [comment]
@renatonascalves can you share more on why you disagree strongly with this one? I think it's fair that if a plugin isn't _at least_ updating to say that it supports recent versions of WordPress and is tested against recent versions of WordPress, it might not be the best idea to have it listed. 

I'm not saying the plugin needs to have a lot of new features, but I think it's fair to say that plugins that want to remain relevant should at least indicate they work with recent versions of WordPress.

**Comment by jasonbahl** [comment]
> Are we using JSON bc long-term we want to serve this via response payload instead of a plugin? Otherwise, using PHP would allow us to just use PHPStan and an array doctype, without all the performance headaches of correctly caching the parsed file.

As discussed over in PR #3188, the JSON will be used as a local file that will be loaded into PHP and then filterable by 3rd party plugins. The JSON will _also_ be used as the source of truth for extensions page on wpgraphql.com/extensions 

So, anytime an extension is added and merged to the .json file, it will be in the extensions page on the website and in the plugin.

**Comment by jasonbahl** [comment]
> If we're anyway adding new workflows, why not drop the PR requirement altogether, and use a GH issue template that the workflow converts to a commit (LT this could be directly be swapped out with a .com form submission > approval process )

I like this idea of a Github Issue template that converts to a commit! Will look into this 👀 

Also will look into a "Submit an Extension" form on the WPGraphQL.com website. 

Possibly both could work 🤔

**Comment by jasonbahl** [comment]
> Not mentioned is whether the plugin is free or where it is hosted. Personally, I recommend making a public repo a requirement (wordpress's svn can count), otherwise we will exacerbate all the review issues @moonmeister highlighted.

I like this idea. We'll only list free, publicly available plugins. Planning to start with Github and .org plugins (there's some UI elements that adjust based on Github or .org), and we can adjust if needed should there be free publicly available plugins hosted elsewhere.

**Comment by renatonascalves** [comment]
@jasonbahl I agree with you. But I also feel there is a history of perfectly working plugins not being updated for any number of reasons. As an example, imagine if WordPress did that in their plugins' listing. Only listing plugins updated based on some criteria. 

But I honestly get the reason here. And since the project plans to introduce more breaking changes soon, this path sort of makes sense.

**Comment by renatonascalves** [comment]
^ that's news to me. But like I said, the path here makes sense.

**Comment by moonmeister** [comment]
There's another issue Gatsby and WP suffer from and that's a million overlapping plugins. I really love plugins that do one thing, and one thing well. Are we planning to do any kind of "curation" on plugins and their quality. I can see a couple scenarios(assume the plugins pass the above criteria:

1. lower quality plugin that mimics something else is submitted
2. Do-it-all plugin submitted that covers things several other plugins already do
3. High quality plugin submitted that could replace an aging less supported plugin even if the old one is still functioning

We also had similar issues with Gatsby Data Source plugins where they weren't implementing the latest and greatest issues and people would fork/re-implement so you'd have 2-3 of the "same" plugin and it was confusing to know which was best. MY work on the team at Gatsby was working to build a "first-class" plugin program. 

To address all these issues I've mentioned I think we accept all plugins that meat the above criteria, but maybe we have "Staff Picks" that help direct folks to the best-in-class plugins for what they do. E.g. The current cors plugin last I checked is a little wanky. But maybe Dovid makes a new one that goes with his plugin and uses all the latest settings toolking we've built so that one gets the ⭐

**Comment by justlevine** [comment]
>  But maybe Dovid makes a new one that goes with his plugin and uses all the latest settings toolking we've built so that one gets the ⭐

https://github.com/AxeWP/wp-graphql-headless-login/pull/129 😬

More seriously though. I think @moonmeister raises a really important point. Curation is necessary to keep a directory useful, but there's both the maintenance of curation and the transparency of the curation process that needs to be considered. E.g. I love the Staff Picks idea but a) who is the staff, and b) if I'm one of the staff i prob shouldn't get to pick.

I _want_ to say that this can be a "lets cross that bridge when we come to it", and we take it iteratively. E.g. start with `Categories` to allow competing plugins to coexist, and go from there.

**Comment by moonmeister** [comment]
All good points. Maybe we're okay with competing plugins within staff picks 🤷‍♂️. I love the idea of getting our elephant involved. Maybe like any good review site we show why we picked one over the other but enable folks to choose for themselves based on our evaluation but we help surface that info. 

Yeah, good call on making this a future aspect. Maybe we create a separate ticket as to not loose track but it's up to you

**Comment by jasonbahl** [comment]
Closing this issue as we now have this documented here: 

- https://github.com/wp-graphql/wp-graphql/blob/develop/docs/submit-extension.md
- https://www.wpgraphql.com/docs/submit-extension

Any specific things that need to be addressed regarding the submission criteria, validation and documentation of said feature should happen in more narrowly scoped issues/prs going forward. 

Thanks all for this conversation and helping polish the extensions page!
